<a href="https://colab.research.google.com/github/renilJoseph/MC_Assignment_Project/blob/master/MC_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# MC Project EEG data classifier

**Objective** : In this notebook we will be training a Neural Network model that will predict the user given their eeg data. 


Here we have the data for 106 subjects with 3 recordings each. Each recording is 2 minutes long and the sensor is capturing data at a frequency of 160Hz. So each recording is a vector of size 19200. Since our data is temporal and can be visualized as a sequence we can use a convolutional neural network. Traditionally CNNs are used for image classification tasks, where it is able to 
figure out spatial relationships between pixels.

More recently CNNs are starting to be used with time series data as well, where they are showing to be very successfull. LSTMs and RNNs are the class of neural networks that are commonly used for temporal data. When we tried our model with LSTM model it have very poor results (<10 % accuracy). CNNs aldo have the added advantage that they can be easily ported to mobiles, as Tensorflow currently does not support exporting LSTM or RNNs to mobile.

**Note**: Currently we do not have any test or validation set since we have very
few data. Once we had a few more data, it is recommended to use a validation set and test set.

**Pre-requisites**

* Scipy [for loading Matlab file]
* Numpy [for numerical computaion]
* Keras - tensorflow [for neural network]

*Authors* : Krishna Babuji, Renil Joseph, Leroy Jacob Vargis, Gopikrishnan Ravindran Nair


## Imports

Here we will import all the necessary libraries for training the model.

In [0]:
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense, Dropout, Conv1D, MaxPooling1D, Flatten
from tensorflow.keras.utils import to_categorical
import tensorflow as tf
import scipy.io

import numpy as np
import pandas as pd

Next  let's download the dataset if it's not already present.

In [0]:
!wget -Nq 'https://raw.githubusercontent.com/renilJoseph/MC_Assignment_Project/master/project/EEGDataset1.mat'

Once we have the data we will transform the data into a pandas dataframe for ease of use.

In [3]:
mat = scipy.io.loadmat('EEGDataset1.mat')
print(f'Mat shape : {mat["Raw_Data"].shape}')

eeg_df1 = pd.DataFrame(mat['Raw_Data'][:, 0, :])
eeg_df1['id'] = eeg_df1.index
eeg_df2 = pd.DataFrame(mat['Raw_Data'][:, 1, :])
eeg_df2['id'] = eeg_df2.index
eeg_df3 = pd.DataFrame(mat['Raw_Data'][:, 2, :])
eeg_df3['id'] = eeg_df3.index

eeg_df = pd.concat([eeg_df1, eeg_df2, eeg_df3], ignore_index=True)
print(f'eeg_df shape : {eeg_df.shape}')

Mat shape : (106, 3, 19200)
eeg_df shape : (318, 19201)


Here we can see that each vector is 19201 long(with the additional Id column).
A reasonable assumption to make is that our predictions won't change much if we change the sampling rate from 160Hz to 80 Hz. Reducing the sampling rate will mean that our LSTM need not handle such a long vector. 

Another option that we have yet to try is just use the first half of the vector, that is use the full sixty seconds with 160 Hz instead of the 120 seconds with 80Hz. But the key idea here is to reduce the dimensions without loosing too much data.

In [0]:
#Down sample from 120Hz to 60Hz
col_drops = [ x for x in eeg_df.columns if x != 'id' and int(x)%2 ==1 ]
eeg_df.drop(col_drops, axis=1, inplace=True)

In [5]:
X_train, y_train = eeg_df.drop(['id'], axis=1), eeg_df['id']
X_train = np.expand_dims(X_train, axis=2)
#X_train = normalize(X_train, axis=0, order=2)
y_train = to_categorical(y_train)
print(X_train.shape, y_train.shape)

(318, 9600, 1) (318, 106)


#Model Definition & Training

Accuracy of the model is not impressive, but we can see that it is better than  a naive model which is random guessing which would be at ~0.01.

If we had more EEG data for these same users we could have created a validation set and tried to tune the hyperparameters more. Since we have only 3 recordings per user, it does not make much sense to create a train and validation set. 

Let's save this model for our predictions.

In [10]:
model = Sequential()
model.add(Conv1D(filters=64, kernel_size=3, activation='relu', input_shape=(9600,1)))
model.add(Conv1D(filters=32, kernel_size=3, activation='relu'))
model.add(Dropout(0.2))
model.add(MaxPooling1D(pool_size=2))
model.add(Flatten())
model.add(Dense(256, activation='relu'))
model.add(Dense(106, activation='softmax'))
model.compile(loss='categorical_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])
print('Train...')

model.fit(X_train, y_train, batch_size=32, shuffle=True, epochs=7)

Train...
Train on 318 samples
Epoch 1/7
318/318 [==============================] - 11s 34ms/sample - loss: 436.9851 - acc: 0.0189
Epoch 2/7
318/318 [==============================] - 10s 33ms/sample - loss: 46.1789 - acc: 0.1038
Epoch 3/7
318/318 [==============================] - 10s 33ms/sample - loss: 5.8665 - acc: 0.1226
Epoch 4/7
318/318 [==============================] - 11s 33ms/sample - loss: 4.2817 - acc: 0.2453
Epoch 5/7
318/318 [==============================] - 11s 33ms/sample - loss: 4.1024 - acc: 0.3836
Epoch 6/7
318/318 [==============================] - 11s 33ms/sample - loss: 3.6798 - acc: 0.5535
Epoch 7/7
318/318 [==============================] - 11s 34ms/sample - loss: 2.6620 - acc: 0.6981


In [0]:
model.save('cnn_eeg.h5')

In [13]:
!toco \
  --output_file=cnn_eeg.tflite\
  --keras_model_file=cnn_eeg.h5

W1206 03:24:54.882520 140497785186176 deprecation.py:506] From /usr/local/lib/python2.7/dist-packages/tensorflow_core/python/ops/init_ops.py:97: calling __init__ (from tensorflow.python.ops.init_ops) with dtype is deprecated and will be removed in a future version.
Instructions for updating:
Call initializer instance with the dtype argument instead of passing it to the constructor
W1206 03:24:54.882921 140497785186176 deprecation.py:506] From /usr/local/lib/python2.7/dist-packages/tensorflow_core/python/ops/init_ops.py:97: calling __init__ (from tensorflow.python.ops.init_ops) with dtype is deprecated and will be removed in a future version.
Instructions for updating:
Call initializer instance with the dtype argument instead of passing it to the constructor
W1206 03:24:54.985518 140497785186176 deprecation.py:506] From /usr/local/lib/python2.7/dist-packages/tensorflow_core/python/ops/resource_variable_ops.py:1630: calling __init__ (from tensorflow.python.ops.resource_variable_ops) with